tools/halide_image_info.h

```diff
@@ -0,0 +1,337 @@
+// This header defines a several methods useful for debugging programs that
+// operate on the Image class supporting images with arbitrary dimensions.
+//
+//   Image<uint16_t> input = load_image(argv[1]);
+//   input.info("input");  // Output the Image header info
+//   input.dump("input");  // Dump the Image data
+//   input.stats("input"); // Collect statistics on the Image
+//
+// These can also be called through macros which automatically tag the
+// output with the symbol name used in the program:
+//
+//   Image_info(input);    // Output the Image header info
+//   Image_dump(input);    // Dump the Image data
+//   Image_stats(input);   // Collect statistics on the Image
+//
+// Info can also be output when an Image is initialized without making any
+// modifications the program by compiling with -DHL_MEMINIT
+//
+// These features are off by default.  To enable, compile with the following
+// flags:
+//
+//   -DHL_MEMINFO  Produce Image info output
+//   -DHL_MEMINIT  Produce Image info when Image is first initialized
+//
+#ifndef HALIDE_TOOLS_IMAGE_INFO_H
+#define HALIDE_TOOLS_IMAGE_INFO_H
+
+#ifndef HL_MEMINFO
+#define Image_info(img)
+#define Image_dump(img)
+#define Image_stats(img)
+#else
+// Stringifying macros to automatically fill in the Image info tag
+#define Image_info(img)  img.info(#img)
+#define Image_dump(img)  img.dump(#img)
+#define Image_stats(img) img.stats(#img)
+#endif // HL_MEMINFO
+
+#if defined(HL_MEMINFO) || defined(HL_MEMINIT)
+#include <cassert>
+#include <cstdlib>
+#include <limits>
+#include <memory>
+#include <iostream>
+#include <iomanip>
+#include <sstream>
+#include <stdint.h>  // <cstdint> requires C++11
+
+#include "HalideRuntime.h"
+
+namespace Halide { namespace Tools {
+
+static inline void print_dimid(int d, int val) {
+    static const char *dimid[] = {"x", "y", "z", "w"};
+    int numdimid = 4;
+    if (d < numdimid)
```

Comment:
Using {} around if and else is pretty much standard style in Halide, even for single-liners.